### PR TITLE
Change PF and RR to expect only backends.

### PR DESCRIPTION
**Pull request by AspirinSJL**
Fix [#12220](https://github.com/grpc/grpc/issues/12220).

### CODE REVIEW

File: src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.c [hunk 1]
```diff
@@ -317,11 +315,11 @@ static void pf_update_locked(grpc_exec_ctx *exec_ctx, grpc_lb_policy *policy,
...
   }
   const grpc_lb_addresses *addresses = arg->value.pointer.p;
-  size_t num_addrs = 0;
   for (size_t i = 0; i < addresses->num_addresses; i++) {
-    if (!addresses->addresses[i].is_balancer) ++num_addrs;
+    // If there were any balancer, we would have chosen grpclb policy instead.
+    GPR_ASSERT(!addresses->addresses[i].is_balancer);
   }
-  if (num_addrs == 0) {
+  if (addresses->num_addresses == 0) {
```
Comment (dgquintas) [inline src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.c]: You can move this `if` up, before the `for`: we don't need to count anything anymore.

File: src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.c [hunk 2]
```diff
@@ -317,11 +315,11 @@ static void pf_update_locked(grpc_exec_ctx *exec_ctx, grpc_lb_policy *policy,
     return;
   }
   const grpc_lb_addresses *addresses = arg->value.pointer.p;
-  size_t num_addrs = 0;
   for (size_t i = 0; i < addresses->num_addresses; i++) {
-    if (!addresses->addresses[i].is_balancer) ++num_addrs;
+    // If there were any balancer, we would have chosen grpclb policy instead.
+    GPR_ASSERT(!addresses->addresses[i].is_balancer);
```
Comment (dgquintas) [inline src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.c]: I'd move this assert to after line 345, right inside the `for` loop: we are already looping there, whereas we don't have any reason to loop here anymore (other than this assert).

File: src/core/ext/filters/client_channel/lb_policy/round_robin/round_robin.c [hunk 3]
```diff
@@ -761,12 +759,13 @@ static void rr_update_locked(grpc_exec_ctx *exec_ctx, grpc_lb_policy *policy,
     return;
   }
   grpc_lb_addresses *addresses = arg->value.pointer.p;
-  size_t num_addrs = 0;
   for (size_t i = 0; i < addresses->num_addresses; i++) {
-    if (!addresses->addresses[i].is_balancer) ++num_addrs;
+    // If there were any balancer, we would have chosen grpclb policy instead.
+    GPR_ASSERT(!addresses->addresses[i].is_balancer);
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/lb_policy/round_robin/round_robin.c]: Same comment here as what David suggested in the PF code: this assertion can be moved to line 800, and this for loop can go away completely.

File: src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.c (hunk 2 above)
Comment (AspirinSJL) [inline src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.c]: Accepted.

File: src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.c (hunk 1 above)
Comment (AspirinSJL) [inline src/core/ext/filters/client_channel/lb_policy/pick_first/pick_first.c]: Included in the previous change.

File: src/core/ext/filters/client_channel/lb_policy/round_robin/round_robin.c (hunk 3 above)
Comment (AspirinSJL) [inline src/core/ext/filters/client_channel/lb_policy/round_robin/round_robin.c]: Accepted.

**Comment by AspirinSJL** [comment]
Got it! Looking forward to my first code merging. Thank you!

**Comment by AspirinSJL** [comment]
retest this please

**Comment by AspirinSJL** [comment]
Asan C++: #12315 
Basic Tests C & C++ Linux [opt] (internal CI): #9150 
Basic Tests Linux (dbg): #9150 
Basic Tests Linux (opt): #9150
Basic Tests Multi-language Linux (internal CI): #12336
Basic Tests Windows: #12337
Interop Cloud-to-Prod Tests (internal CI): #10763
Msan C (internal CI): #12338

**Pull request merged**